### PR TITLE
Heterogeneous output in Lambdify

### PR DESCRIPTION
**Pull request by bjodah**
To address #107.

There are some test failures still. I'll update and ping once this is ready for review.
I changed my mind and this now only returns a tuple if multiple expressions are given (e.g. a vector and a matrix). This means that this should be a non-breaking change.

### CODE REVIEW

**Comment by isuruf** [comment]
@bjodah, Lambdify's `__call__` has considerable overhead.

Here's some python code that I wrote,
```Python

outputs_ravel = list(itertools.chain(*[output.tolist() for output in outputs]))
cb = symengine.Lambdify(inputs, outputs_ravel, backend="llvm")

def func(*args):
    result = []
    n = numpy.empty(len(outputs_ravel))
    t = cb.unsafe_real(np.concatenate([arg.ravel() for arg in args]), n)
    start = 0
    for output in outputs:
        elems = reduce(mul, output.shape)
        result.append(n[start:start+elems].reshape(output.shape))
        start += elems
    return result
    
return lambda *args: func(*args)
```
This makes SymEngine as fast as PyDy's cythonize. Updated the timings here, https://github.com/symengine/symengine/pull/1094#issuecomment-254212768

**Comment by bjodah** [comment]
@isuruf 
Thanks, I'll need to make the tests pass and then I'll try to optimize `Lambdify.__call__`.
One approach would be to make NumPy a hard dependency of `symengine.py`, how would you feel about that? `cdef`'ing arguments as `cnp.ndarray` will most likely remove a large portion of the overhead - constructing memoryviews is quite expensive, and dealing with nested python-lists/tuples even more so. The code around `Lambdify` would also be considerably simpler.

**Comment by bjodah** [comment]
@isuruf I added the `-a` flag to cython and it looks like the overhead in `__call__` is related to the use
of memoryviews. Should we make NumPy a hard dependency of Lambdify? (that would make the code less complex and considerably faster). I could of course code a "fast path" for NumPy but I feel like it's getting out of hand already trying to support Cython's array, Python's array.array and NumPy arrays.

**Comment by isuruf** [comment]
I'm fine with making numpy a hard dependency of just Lambdify

**Comment by isuruf** [comment]
@bjodah, note that in https://github.com/symengine/symengine.py/pull/112#issuecomment-259082718 I've used unsafe_real and it has memoryviews in it, so there might be another issue as well.

**Comment by bjodah** [comment]
I'm looking into using SymEngine in the codegen tutorial at SciPy.
So I started looking into this branch again.

But when I try to compile symengine with llvm I get a test failure:
```
33/37 Test #33: test_lambda_double ...............***Failed    0.00 sec

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
test_lambda_double is a Catch v1.1 b14 (develop) host application.
Run with -? for options

-------------------------------------------------------------------------------
Check llvm and lambda are equal
-------------------------------------------------------------------------------
/opt/symengine/symengine/tests/eval/test_lambda_double.cpp:173
...............................................................................

/opt/symengine/symengine/tests/eval/test_lambda_double.cpp:173: FAILED:
due to a fatal error condition:
```
I have created this Dockerized environment which reproduces the error:
https://github.com/bjodah/gsoc2017-bjodah/tree/master/se_llvm

I tried to follow the Travis scripts closely but obviously I failed. Any idea what I am doing wrong?

**Comment by isuruf** [comment]
I think this is because llvm was compiled with gcc 4.8.5 and you are using gcc 5.4.0. I'm surprised that there were no link errors.

**Comment by bjodah** [comment]
could be, modifying CPPFLAGS did not help:
```
$ git diff
diff --git a/se_llvm/20_compile_symengine.sh b/se_llvm/20_compile_symengine.sh
index e116ece..a3545d6 100755
--- a/se_llvm/20_compile_symengine.sh
+++ b/se_llvm/20_compile_symengine.sh
@@ -11,6 +11,7 @@ export CXX="ccache clang++"
 export CC="ccache clang"
 export CCACHE_DIR=/ccache
 ccache -M 400M
+export CPPFLAGS="-D_GLIBCXX_USE_CXX11_ABI=0"
 cmake -DCMAKE_INSTALL_PREFIX=$our_install_dir -DCMAKE_BUILD_TYPE:STRING="Release" -DWITH_LLVM:BOOL=ON -DINTEGER_CLASS:STRING=gmp -DBUILD_SHARED_LIBS:BOOL=ON -DWITH_MPC=yes $SOURCE_DIR
 make -j 4
 make install
```
gives same error. I thought setting CXX and CC was enough (I set them to use clang with ccache):
```
+-------------------------------+
| Configuration results SUMMARY |
+-------------------------------+

CMAKE_C_COMPILER:   /opt/se/bin/ccache
CMAKE_CXX_COMPILER: /opt/se/bin/ccache
CMAKE_BUILD_TYPE: Release
...
```

**Comment by bjodah** [comment]
Current status of `benchmark/Lambdify.py`:
```
$ CFLAGS="-march=native -ffast-math -O3" python benchmarks/Lambdify.py 
SymEngine (lambda double) speed-up factor (higher is better) vs sympy:       5.1172
symengine (LLVM)          speed-up factor (higher is better) vs sympy:       5.7401
symengine (ManualLLVM)    speed-up factor (higher is better) vs sympy:       6.4252
benchmarks/Lambdify.py:78: UserWarning: Cython code for Lambdify.__call__ is slow.
  warnings.warn("Cython code for Lambdify.__call__ is slow.")
Hard-coded Cython code    speed-up factor (higher is better) vs sympy:       23.096
```
(speed-up factors below 1 before moving `use_numpy` kwarg from `__call__` to `__init__`. I am now investigating numpy array vs. memoryviews.

**Comment by isuruf** [comment]
@bjodah, can you post the timings without `CFLAGS="-march=native -ffast-math -O3"` ? We haven't turned on aggressive optimizations in LLVM.

**Comment by bjodah** [comment]
Sure, that was a factor of ~11

**Comment by bjodah** [comment]
Hmm... those numbers were against SymPy master. Using sympy-1.0 I now get:
```
Hard-coded Cython code          speed-up factor (higher is better) vs sympy:       8.7338
```
so apparently SymPy master's lambdify has had a regression (1/3 of the performance).

**Comment by bjodah** [comment]
Comparing against sympy-1.0 & using -O1:
```
$ CFLAGS="-O1" python benchmarks/Lambdify.py 
SymEngine (lambda double)       speed-up factor (higher is better) vs sympy:       1.9232
symengine (lambda double + CSE) speed-up factor (higher is better) vs sympy:      0.98464
symengine (LLVM)                speed-up factor (higher is better) vs sympy:       2.3853
symengine (ManualLLVM)          speed-up factor (higher is better) vs sympy:       2.2815
Hard-coded Cython code          speed-up factor (higher is better) vs sympy:       3.8374
```

**Comment by bjodah** [comment]
I think making NumPy a hard requirement for `Lambdify` has a some trade-offs:
- slightly better performance
- arguably more readable code (source of `Lambdify.__call__` more accessible now)
- a bit more tricky setup.py -- we need some compile option `WITH_NUMP` or so

on that last point: I have to implement that still. What is the best approach for that? (if we want to make NumPy i hard requirement that is). I'm thinking putting the whole Lambdify definition inside a block?, i.e.
```
IF WITH_NUMPY:
    cdef class Lambdify:
        ...
```

**Comment by bjodah** [comment]
(regarding the `sympy.lambdify` "regression": there was none, I just needed to specify `modules='math'`)

**Comment by isuruf** [comment]
If you have some time, can you run this example, https://github.com/pydy/pydy/pull/360 ? If not, I'll be able to run it over the weekend.

**Comment by isuruf** [comment]
> putting the whole Lambdify definition inside a block

I'm not sure if `cimport`s work inside such blocks. If it does, then that's probably the best approach

**Comment by bjodah** [comment]
Sure - I'll run those later tonight or tomorrow.

**Comment by bjodah** [comment]
OK, so I've spent some time trying to run those benchmarks using python3 and:
pydy@c1239b4
sympy@39f7bff
symengine.py@ca777f6
symengine@v0.3.0

But I get a traceback:
```
Pendulum with 1 links.
======================
Traceback (most recent call last):
  File "bin/benchmark_pydy_code_gen.py", line 126, in <module>
    run_benchmark(args.max_num_links, args.num_time_steps)
  File "bin/benchmark_pydy_code_gen.py", line 38, in run_benchmark
    sys = n_link_pendulum_on_cart(n, cart_force=False)
  File "/home/bjorn/vc/pydy/pydy/models.py", line 234, in n_link_pendulum_on_cart
    kane.kanes_equations(particles, forces)
  File "/home/bjorn/vc/sympy/sympy/physics/mechanics/kane.py", line 724, in kanes_equations
    fr = self._form_fr(loads)
  File "/home/bjorn/vc/sympy/sympy/physics/mechanics/kane.py", line 278, in _form_fr
    vel_list = [msubs(i, self._qdot_u_map) for i in vel_list]
  File "/home/bjorn/vc/sympy/sympy/physics/mechanics/kane.py", line 278, in <listcomp>
    vel_list = [msubs(i, self._qdot_u_map) for i in vel_list]
  File "/home/bjorn/vc/sympy/sympy/physics/mechanics/functions.py", line 469, in msubs
    return expr.applyfunc(lambda x: func(x, sub_dict))
  File "/home/bjorn/vc/sympy/sympy/physics/vector/vector.py", line 693, in applyfunc
    ov += Vector([(v[0].applyfunc(f), v[1])])
  File "symengine/lib/symengine_wrapper.pyx", line 2223, in symengine.lib.symengine_wrapper.DenseMatrixBase.applyfunc (/home/bjorn/vc/symengine.py/symengine/lib/symengine_wrapper.cpp:63523)
  File "/home/bjorn/vc/sympy/sympy/physics/mechanics/functions.py", line 469, in <lambda>
    return expr.applyfunc(lambda x: func(x, sub_dict))
  File "/home/bjorn/vc/sympy/sympy/physics/mechanics/functions.py", line 467, in <lambda>
    func = lambda expr, sub_dict: _crawl(expr, _sub_func, sub_dict)
  File "/home/bjorn/vc/sympy/sympy/physics/mechanics/functions.py", line 480, in _crawl
    return expr.func(*new_args)
TypeError: module.__init__() argument 1 must be str, not Symbol
```
So there seems to be some fixing needed, but I'm not familiar with `pydy` nor `sympy.physics.mechanics` I'm afraid, maybe we could add a self-contained benchmark to to the `symengine.py/benchmarks/` folder? (that way the benchmarks could be tested during CI runs to ensure they keep working?)

Regarding my last 2 commits: I'm quite new to writing CMake-scripts so there might be things to improve there.

**Comment by isuruf** [comment]
Yes, some changes in SymPy breaks the benchmark. Can you aply the following diff for sympy for now?
```diff
diff --git a/sympy/physics/mechanics/functions.py b/sympy/physics/mechanics/functions.py
index 2aea3bb..14adc78 100644
--- a/sympy/physics/mechanics/functions.py
+++ b/sympy/physics/mechanics/functions.py
@@ -10,7 +10,7 @@
 from sympy.physics.mechanics.rigidbody import RigidBody
 from sympy import simplify
 from sympy.core.backend import (Matrix, sympify, Mul, Derivative, sin, cos,
-                                tan, AppliedUndef, S)
+                                tan, AppliedUndef, S, USE_SYMENGINE)
 
 __all__ = ['inertia',
            'inertia_of_point_mass',
@@ -467,8 +467,8 @@ def msubs(expr, *sub_dicts, **kwargs):
     smart = kwargs.pop('smart', False)
     if smart:
         func = _smart_subs
-    elif hasattr(expr, 'msubs'):
-        return expr.msubs(sub_dict)
+    elif USE_SYMENGINE:
+        func = lambda expr, sub_dict: expr.msubs(sub_dict)
     else:
         func = lambda expr, sub_dict: _crawl(expr, _sub_func, sub_dict)
     if isinstance(expr, (Matrix, Vector, Dyadic)):
```

**Comment by bjodah** [comment]
For `links = 10` and `time_steps=1000` using SymEngine:
```
The derivation took 0.18646 seconds.

Generating with lambdify method.
--------------------------------
The code generation took 2.79032 seconds.
ODE integration took 4.79481 seconds.
```
not using SymEngine (note that setting USE_SYMENGINE=0 causes sympy to use symengine...):
```
The derivation took 3.05501 seconds.

Generating with lambdify method.
--------------------------------
The code generation took 0.81127 seconds.
ODE integration took 390.68950 seconds.

Generating with cython method.
------------------------------
The code generation took 69.66981 seconds.
ODE integration took 4.25866 seconds.
```

**Comment by isuruf** [comment]
I ran the benchmark with master, PR + ManualLLVM, PR + LLVM
Used the following patch to use this PRs multiple output feature.
```
@@ -682,24 +688,9 @@ class LambdifyODEFunctionGenerator(ODEFunctionGenerator):
 
         if USE_SYMENGINE:
             import symengine
-            outputs_ravel = list(chain(*outputs))
-            try:
-                cb = symengine.Lambdify(inputs, outputs_ravel, backend="llvm")
-            except (TypeError, ValueError):
-                # TypeError if symengine is old, ValueError if symengine is
-                # not compiled with llvm support
-                cb = symengine.Lambdify(inputs, outputs_ravel)
+            cb = symengine.Lambdify(inputs, *outputs)
             def func(*args):
-                result = []
-                n = np.empty(len(outputs_ravel))
-                cb.unsafe_real(np.concatenate([a.ravel() for a in args]), n)
-                start = 0
-                for output in outputs:
-                    elems = reduce(mul, output.shape)
-                    result.append(n[start : (start + elems)]
-                                        .reshape(output.shape))
-                    start += elems
-                return result
+                return cb(np.concatenate([a.ravel() for a in args]))
             return func
         else:
             modules = [{'ImmutableMatrix': np.array}, 'numpy']
```

ManualLLVM is still faster,

```

**Comment by isuruf** [comment]
Oops, pressed enter before finishing commenting.

```
Pendulum with 10 links.
=======================
The derivation took 0.16069 seconds.

Generating with lambdify method.
--------------------------------
The code generation took 0.19171 seconds.
ODE integration took 5.25970 seconds.
```

```
Pendulum with 10 links.
=======================
The derivation took 0.15894 seconds.

Generating with lambdify method.
--------------------------------
The code generation took 0.19862 seconds.
ODE integration took 4.90780 seconds.
```

Compared to master, this PR is better.
```
Pendulum with 10 links.
=======================
The derivation took 0.16067 seconds.

Generating with lambdify method.
--------------------------------
The code generation took 0.19473 seconds.
ODE integration took 5.08710 seconds.
```

**Comment by bjodah** [comment]
OK, so I extracted a small test case reproducing ManualLLVM being faster:
```
$ python3 benchmarks/Lambdify_6_links_rhs.py 
[1, 1, 1, 1, 1, 1, 1, 21*cos(1), -6*cos(1), -5*cos(1), -4*cos(1), -3*cos(1), -2*cos(1), -cos(1)]
SymEngine (lambda double)       speed-up factor (higher is better) vs sympy:       2.6123
symengine (LLVM)                speed-up factor (higher is better) vs sympy:       8.1947
symengine (ManualLLVM)          speed-up factor (higher is better) vs sympy:       8.4328
benchmarks/Lambdify_6_links_rhs.py:94: UserWarning: Cython code for Lambdify.__call__ is slow.
  warnings.warn("Cython code for Lambdify.__call__ is slow.")
Hard-coded Cython code          speed-up factor (higher is better) vs sympy:        25.12
```
Will work on improving that benchmark.

**Comment by isuruf** [comment]
That benchmark also has one output. We should get a benchmark for heterogeneous output in Lambdify.

**Comment by bjodah** [comment]
you're right. I'll update the benchmark (tomorrow or Monday) to also calculate: f(t), the Jacobian matrix of f(t) and df/dt in the same call.

**Comment by bjodah** [comment]
`__call__` is now running slightly faster than `ManualLLVM` for heterogeneous output:
```
python3 benchmarks/heterogenous_output_Lambdify.py
SymEngine (lambda double)       speed-up factor (higher is better) vs sympy:       40.702
symengine (LLVM)                speed-up factor (higher is better) vs sympy:       248.95
symengine (ManualLLVM)          speed-up factor (higher is better) vs sympy:       244.31
```

File: symengine/lib/symengine_wrapper.pyx [hunk 1]
```diff
@@ -2996,402 +3001,401 @@ def powermod_list(a, b, m):
...
+        """
+        cdef size_t args_size, tot_out_size
+        cdef list out_shapes
+        cdef readonly bint real
+        cdef readonly int n_exprs
+        cdef int *out_sizes
+        cdef int *accum_out_sizes
+        cdef object numpy_dtype
+
+        def __cinit__(self, args, *exprs, bool real=True):
```
Comment (isuruf) [inline symengine/lib/symengine_wrapper.pyx]: Keyword-only arguments are python 3 only right?

Comment (bjodah) [inline symengine/lib/symengine_wrapper.pyx]: I believe keyword only args are when `**kwargs` is also present.

Comment (isuruf) [inline symengine/lib/symengine_wrapper.pyx]: See https://www.python.org/dev/peps/pep-3102/

```
Syntactically, the proposed changes are fairly simple. The first change is to allow regular arguments to appear after a varargs argument:

def sortwords(*wordlist, case_sensitive=False):
    ...
```

Comment (bjodah) [inline symengine/lib/symengine_wrapper.pyx]: Ah, I see, but Cython supports it even for Python 2 right?

**Comment by bjodah** [comment]
@isuruf I think this might be ready for review

File: CMakeLists.txt [hunk 2]
```diff
@@ -59,6 +58,14 @@ else()
     set(HAVE_SYMENGINE_LLVM False)
 endif()
 
+if(WITH_NUMPY)
+    find_package(NumPy REQUIRED)
+    include_directories(${NUMPY_INCLUDE_PATH})
```
Comment (isuruf) [inline CMakeLists.txt]: Are we not linking the numpy libraries?

File: symengine/lib/symengine_wrapper.pyx [hunk 3]
```diff
@@ -2996,402 +3001,402 @@ def powermod_list(a, b, m):
...
-            try:
-                import numpy as np
-            except ImportError:
-                use_numpy = False  # we will use cython.view.array instead
+        """
+        cdef size_t args_size, tot_out_size
+        cdef list out_shapes
+        cdef readonly bint real
+        cdef readonly int n_exprs
+        cdef int *out_sizes
```
Comment (isuruf) [inline symengine/lib/symengine_wrapper.pyx]: Can this be a std::vector?

File: CMakeLists.txt (hunk 2 above)
Comment (isuruf) [inline CMakeLists.txt]: I just checked and there are no symbols from numpy libraries linked in. It seems we are not using anything that requires linking. I wonder whether that means we can build with one numpy version and have it work on other numpy versions.

Comment (bjodah) [inline CMakeLists.txt]: Yes, it is my understanding that the NumPy C API is largely back-compatible. And no, the numpy functions used in the code uses the python layer (e.g. numpy.empty). It is possible that we could gain some performance by directly calling the NumPy C API in `__call__` (that would then require linking).

Comment (isuruf) [inline CMakeLists.txt]: If so, I think we can just use `numpy.ndarray.ctypes` to get a array pointer and have numpy only as a runtime requirement. I'll have a look to see if this is possible.

**Comment by certik** [comment]
To make it clear --- `symengine.py` belongs to the scipy stack. `symengine` itself should only depend on a few well maintained C++ libraries.

**Comment by isuruf** [comment]
Thanks for the PR, @bjodah

**Comment by isuruf** [comment]
@bjodah, packaging symengine for conda with numpy as a build-time dependency means the conda package has to be compiled for each numpy version. I'd like to avoid that if possible with numpy only as a runtime requirement. Can you try my branch at https://github.com/isuruf/symengine.py/tree/memview and see if memviews have a performance penalty? I ran the benchmarks myself and for me I don't see much difference.

**Comment by isuruf** [comment]
ping, @bjodah

**Comment by bjodah** [comment]
@isuruf just tried it. No the performance is the same on my machine. +1 for using memviews

**Pull request merged**